### PR TITLE
Add sled as an alternative storage backend

### PR DESCRIPTION
**Pull request by pvdrz**
This adds [sled](https://sled.rs) as an alternative storage backend.

~~Blocked on https://github.com/iotaledger/bee/pull/599 due to the changes in the `AsStream` trait.~~

### CODE REVIEW

**Review by louaykamel — COMMENTED** [review]
I reviewed the files and it LGTM. 
However will do some tests early morning.

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
`// Copyright 2021 IOTA Stiftung` for the new files :)

File: bee-storage/bee-storage-sled/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+[package]
+name = "bee-storage-sled"
+version = "0.1.0"
+authors = ["IOTA Stiftung"]
+edition = "2018"
+description = "A bee-storage implementation for the Sled backend"
+readme = "README.md"
+repository = "https://github.com/iotaledger/bee"
+license = "Apache-2.0"
+keywords = ["iota", "tangle", "bee", "framework", "storage", "sled"]
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-sled/Cargo.toml]: Maximum 5 keywords to publish a crate on crates.io. Yes I know rocksdb has 6, I just noticed it, and it's not published yet. Maybe remove `storage`. Also do the same for the rocksdb crate ?

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: bee-storage/bee-storage-sled/Cargo.toml (hunk 1 above)
Comment (thibault-martinez) [reply bee-storage/bee-storage-sled/Cargo.toml]: Not solved ?

Comment (pvdrz) [reply bee-storage/bee-storage-sled/Cargo.toml]: eventual consistency

File: bee-storage/bee-storage-sled/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-sled/src/lib.rs:3]:
```
#![deny(missing_docs)]
#![deny(warnings)]
```

Comment (thibault-martinez) [reply bee-storage/bee-storage-sled/src/lib.rs:3]: I noticed we missed it for the rocks crate, I will fix it

**Review by thibault-martinez — CHANGES_REQUESTED** [review]
Can you add an entry in the CHANGELOG of sled please ?

File: bee-storage/bee-storage-sled/src/storage.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    async fn shutdown(self) -> Result<(), Self::Error> {
+        self.inner.flush()?;
+        Ok(())
+    }
+
+    async fn size(&self) -> Result<Option<usize>, Self::Error> {
+        Ok(Some(self.inner.size_on_disk()? as usize))
+    }
+
+    async fn get_health(&self) -> Result<Option<StorageHealth>, Self::Error> {
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-sled/src/storage.rs:104]: Health needs to be handled

File: bee-storage/bee-storage-sled/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+bee-message = { version = "0.1.3", path = "../../bee-message" }
+bee-storage = { version = "0.5.0", path = "../bee-storage" }
+bee-tangle = { version = "0.1.0", path = "../../bee-tangle" }
+
+async-trait = "0.1"
+futures = "0.3"
+num_cpus = "1.13"
+pin-project = "1.0"
+sled = { version = "0.34.6", features = [ "compression" ]}
+serde = { version = "1.0", features = [ "derive" ] }
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-sled/Cargo.toml:24]: serde before sled

File: bee-storage/bee-storage-sled/src/trees.rs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+pub const TREE_MESSAGE_ID_TO_MESSAGE_ID: &str = "message_id_to_message_id";
+/// Identifier for the `PaddedIndex` to `MessageId` tree.
+pub const TREE_INDEX_TO_MESSAGE_ID: &str = "index_to_message_id";
+/// Identifier for the `OutputId` to `CreatedOutput` tree.
+pub const TREE_OUTPUT_ID_TO_CREATED_OUTPUT: &str = "output_id_to_created_output";
+/// Identifier for the `OutputId` to `ConsumedOutput` tree.
+pub const TREE_OUTPUT_ID_TO_CONSUMED_OUTPUT: &str = "output_id_to_consumed_output";
+/// Identifier for the `Unspent` tree.
+pub const TREE_OUTPUT_ID_UNSPENT: &str = "output_id_unspent";
+/// Identifier for the `Ed25519` to `OutputId` tree.
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-sled/src/trees.rs]:
```suggestion
/// Identifier for the `Ed25519Address` to `Vec<OutputId>` tree.
```

File: bee-storage/bee-storage-sled/src/trees.rs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+//! Identifiers for each tree.
+//!
+//! Sled allows creating new, isolated keyspaces by adding new trees to the database. Each tree can
+//! be accessed using the `sled::Db::open_tree` method with one of the identifiers found here.
+
+/// Identifier for the `MessageId` to `Message` tree.
+pub const TREE_MESSAGE_ID_TO_MESSAGE: &str = "message_id_to_message";
+/// Identifier for the `MessageId` to `MessageMetadata` tree.
+pub const TREE_MESSAGE_ID_TO_METADATA: &str = "message_id_to_metadata";
+/// Identifier for the `MessageId` to `MessageId` tree.
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-sled/src/trees.rs]:
```suggestion
/// Identifier for the `MessageId` to `Vec<MessageId>` tree.
```

File: bee-storage/bee-storage-sled/src/trees.rs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+pub const TREE_MESSAGE_ID_TO_MESSAGE_ID: &str = "message_id_to_message_id";
+/// Identifier for the `PaddedIndex` to `MessageId` tree.
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-sled/src/trees.rs]:
```suggestion
/// Identifier for the `PaddedIndex` to `Vec<MessageId>` tree.
```

File: bee-storage/bee-storage-sled/src/trees.rs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+pub const TREE_SOLID_ENTRY_POINT_TO_MILESTONE_INDEX: &str = "solid_entry_point_to_milestone_index";
+/// Identifier for the `MilestoneIndex` to `OutputDiff` tree.
+pub const TREE_MILESTONE_INDEX_TO_OUTPUT_DIFF: &str = "milestone_index_to_output_diff";
+/// Identifier for the `Address` to `Balance` tree.
+pub const TREE_ADDRESS_TO_BALANCE: &str = "address_to_balance";
+/// Identifier for the `MilestoneIndex` to `Vec<UnreferencedMessage>` tree.
+pub const TREE_MILESTONE_INDEX_TO_UNREFERENCED_MESSAGE: &str = "milestone_index_to_unreferenced_message";
+/// Identifier for the `MilestoneIndex` to `Vec<Receipt>` tree.
+pub const TREE_MILESTONE_INDEX_TO_RECEIPT: &str = "milestone_index_to_receipt";
+/// Identifier for the `true` to `Vec<TreasuryOutput>` tree.
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-sled/src/trees.rs]:
```suggestion
/// Identifier for the `bool` to `Vec<TreasuryOutput>` tree.
```

File: bee-storage/bee-storage-sled/tests/access.rs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+#[macro_use]
+#[allow(unused_macros)]
+macro_rules! impl_access_test {
+    ($name:ident) => {
+        #[tokio::test]
+        async fn $name() {
+            use bee_storage::backend::StorageBackend;
+
+            let path = String::from("./tests/database/") + stringify!($name);
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-sled/tests/access.rs:12]: Can you gitignore this folder ? Also the rocksdb one.

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: bee-storage/bee-storage-sled/src/access/fetch.rs [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+};
+
+use std::convert::{TryFrom, TryInto};
+
+#[async_trait::async_trait]
+impl Fetch<u8, System> for Storage {
+    async fn fetch(&self, key: &u8) -> Result<Option<System>, <Self as StorageBackend>::Error> {
+        Ok(self
+            .inner
+            .get(&[*key])?
```
Comment (Adam-Gleave) [inline bee-storage/bee-storage-sled/src/access/fetch.rs]: Maybe expand this? It takes (me) two seconds to realise what this is doing.

File: bee-storage/bee-storage/src/lib.rs [hunk 11]
```diff
@@ -20,3 +20,4 @@
 pub mod access;
 pub mod backend;
 pub mod health;
```
Comment (thibault-martinez) [inline bee-storage/bee-storage/src/lib.rs]: Maybe then put health in system together with version ?

File: bee-storage/bee-storage-sled/src/storage.rs [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    /// There is a storage version mismatch between the storage folder and this version of the
+    /// storage.
+    #[error("Storage version mismatch, {0:?} != {1:?}, remove storage folder and restart")]
+    VersionMismatch(StorageVersion, StorageVersion),
+    /// The storage was not closed properly.
+    #[error("Unhealthy storage: {0:?}, remove storage folder and restart")]
+    UnhealthyStorage(StorageHealth),
+}
+
+pub(crate) const STORAGE_VERSION: StorageVersion = StorageVersion(1);
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-sled/src/storage.rs]:
```suggestion
pub(crate) const STORAGE_VERSION: StorageVersion = StorageVersion(0);
```

File: bee-storage/bee-storage/src/system/mod.rs [hunk 13]
```diff
@@ -1,29 +1,41 @@
...
+/// Errors to be raised if packing/unpacking `System` fails.
 #[derive(Debug, thiserror::Error)]
 pub enum Error {
+    /// I/O error.
     #[error("I/O error happened: {0}")]
     Io(#[from] std::io::Error),
+    /// Packing/unpacking the `System::Health` variant failed.
     #[error("Storage health error: {0}")]
     Health(#[from] StorageHealthError),
+    /// Found and invalid key while unpacking a `System` value.
```
Comment (thibault-martinez) [inline bee-storage/bee-storage/src/system/mod.rs]:
```suggestion
    /// Found an invalid key while unpacking a `System` value.
```

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: bee-storage/bee-storage/src/lib.rs (hunk 11 above)
Comment (pvdrz) [reply bee-storage/bee-storage/src/lib.rs]: makes sense

File: bee-storage/bee-storage-sled/src/storage.rs [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    access::{Fetch, Insert},
+    backend::StorageBackend,
+    system::{StorageHealth, StorageVersion, System, SYSTEM_HEALTH_KEY, SYSTEM_VERSION_KEY},
+};
+
+use async_trait::async_trait;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+/// Error to be raised when a backend operation fails.
```
Comment (thibault-martinez) [inline bee-storage/bee-storage-sled/src/storage.rs:17]: Comment before attributes

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Review by thibault-martinez — DISMISSED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**